### PR TITLE
Fix groupby().count() for datetime columns

### LINKED ISSUE #13393: Pandas 0.18.1 df.groupby().count() throws "ValueError: Buffer has wrong number of dimensions" when one of the counted columns has dtype datetime64
I've confirmed the error does not occur in Pandas 0.16 on a similar machine.
#### Code Sample, a copy-pastable example if possible

```
import pandas as pd
df = pd.DataFrame({'x': ['a', 'a', 'b'],
                   'y': [pd.Timestamp('2016-05-07 20:09:25+00:00'), pd.Timestamp('2016-05-07 20:09:29+00:00'), pd.Timestamp('2016-05-07 20:09:29+00:00')]})
df.groupby('x').count()
```
#### Observed output

---

```
ValueError                                Traceback (most recent call last)
<ipython-input-5-3119045de5b1> in <module>()
      2 df = pd.DataFrame({'x': ['a', 'a', 'b'],
      3                    'y': [pd.Timestamp('2016-05-07 20:09:25+00:00'), pd.Timestamp('2016-05-07 20:09:29+00:00'), pd.Timestamp('2016-05-07 20:09:29+00:00')]})
----> 4 print df.groupby('x').count()

/usr/local/lib/python2.7/dist-packages/pandas/core/groupby.pyc in count(self)
   3754         blk = map(make_block, map(counter, val), loc)
   3755 
-> 3756         return self._wrap_agged_blocks(data.items, list(blk))
   3757 
   3758 

pandas/lib.pyx in pandas.lib.count_level_2d (pandas/lib.c:23068)()

ValueError: Buffer has wrong number of dimensions (expected 2, got 1)
```
#### Expected Output

```
x  y
a  2
b  1
```
#### output of `pd.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-32-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.1
pip: 8.1.2
setuptools: 21.2.1
Cython: None
numpy: 1.11.0
scipy: 0.16.0
statsmodels: 0.5.0
xarray: None
IPython: 4.2.0
sphinx: None
patsy: 0.2.1
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: 0.8.0
tables: 3.1.1
numexpr: 2.2.2
matplotlib: 1.5.1
openpyxl: 1.7.0
xlrd: 0.9.2
xlwt: 0.7.5
xlsxwriter: 0.5.8
lxml: None
bs4: 4.2.1
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 0.9.7
pymysql: None
psycopg2: 2.5.3 (dt dec pq3 ext)
jinja2: 2.8
boto: 2.36.0
pandas_datareader: None
```

### PR DESCRIPTION
**Pull request by watercrossing**
- [x] closes #13393
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by watercrossing** [comment]
In which whatsnew file should the whatsnew entry go?

File: pandas/tests/groupby/test_counting.py [hunk 1]
```diff
@@ -195,3 +195,13 @@ def test_ngroup_respects_groupby_order(self):
                                 g.ngroup())
             assert_series_equal(Series(df['group_index'].values),
                                 g.cumcount())
+
+    def test_count_with_datetime(self):
+        df = DataFrame({'x': ['a', 'a', 'b'],
```
Comment (jreback) [inline pandas/tests/groupby/test_counting.py]: add gh issue number as a comment

File: pandas/core/groupby.py [hunk 2]
```diff
@@ -4365,7 +4365,10 @@ def count(self):
         ids, _, ngroups = self.grouper.group_info
         mask = ids != -1
 
-        val = ((mask & ~isna(blk.get_values())) for blk in data.blocks)
+        makeNDarray = lambda vals: vals[None, :] if vals.ndim == 1 else vals
+
+        val = ((mask & ~isna(makeNDarray(blk.get_values())))
```
Comment (jreback) [inline pandas/core/groupby.py]: just use np.atleast_2d
no need for separate mask

**Review by jreback — CHANGES_REQUESTED** [review]
you can put in 0.21.1 bug fixes

File: pandas/tests/groupby/test_counting.py [hunk 3]
```diff
@@ -195,3 +195,14 @@ def test_ngroup_respects_groupby_order(self):
                                 g.ngroup())
             assert_series_equal(Series(df['group_index'].values),
                                 g.cumcount())
+
+    def test_count_with_datetime(self):
```
Comment (jreback) [inline pandas/tests/groupby/test_counting.py]: can u parametrize to count over tz naive (and aware);
also over Timedeltas and Period would be great

Comment (jreback) [reply pandas/tests/groupby/test_counting.py]: change test name to _datetimelike

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_counting.py [hunk 4]
```diff
@@ -195,3 +196,28 @@ def test_ngroup_respects_groupby_order(self):
...
             assert_series_equal(Series(df['group_index'].values),
                                 g.cumcount())
+
+    def test_count_with_datetimelike(self):
+        # test for #13393, where DataframeGroupBy.count() fails
+        # when counting a datetimelike column.
+
+        expected = DataFrame({'y': [2, 1]}, index=['a', 'b'])
+        expected.index.name = "x"
+
```
Comment (jreback) [inline pandas/tests/groupby/test_counting.py]: use parametrize (the decorator)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_counting.py [hunk 5]
```diff
@@ -195,3 +196,28 @@ def test_ngroup_respects_groupby_order(self):
...
+            df = DataFrame({'x': ['a', 'a', 'b'], 'y': datetimelike})
+            res = df.groupby('x').count()
+            assert_frame_equal(expected, res)
+
+        d1 = [Timestamp('2016-05-07 20:09:25+00:00'),
+              Timestamp('2016-05-07 20:09:29+00:00'),
+              Timestamp('2016-05-07 20:09:29+00:00')]
+        _test(d1)
+        d2 = [Timedelta(x, unit="h") for x in range(1, 4)]
+        _test(d2)
```
Comment (jreback) [inline pandas/tests/groupby/test_counting.py]: you need to change Timedelta and Period to work here

File: pandas/tests/groupby/test_counting.py (hunk 4 above)
Comment (watercrossing) [reply pandas/tests/groupby/test_counting.py]: Oh! I looked in the wrong test file for inspiration then.

File: pandas/tests/groupby/test_counting.py (hunk 5 above)
Comment (watercrossing) [reply pandas/tests/groupby/test_counting.py]: I am not sure I understand your request.

Comment (jreback) [reply pandas/tests/groupby/test_counting.py]: the examples for timedelta and period are unique and won’t group to the same as the other examples 

just change to individually construct them like you did for time stamps

Comment (watercrossing) [reply pandas/tests/groupby/test_counting.py]: but we are not grouping by the timestamps, it doesn't matter if they are unique or not; the original issue fails regardless.

**Comment by watercrossing** [comment]
I don't think the Travis Issue is related.

**Comment by jreback** [comment]
thanks @watercrossing, nice PR!

**Pull request merged**